merkle_tree/src/internal.rs

```diff
@@ -131,52 +128,76 @@ impl<T: NodeValue> MerkleCommitment<T> for MerkleTreeCommitment<T> {
     }
 }
 
-/// Merkle proof struct.
-#[derive(Derivative, Debug, Clone, Serialize, Deserialize)]
-#[derivative(Eq, Hash, PartialEq)]
-#[serde(bound = "E: CanonicalSerialize + CanonicalDeserialize,
-             I: CanonicalSerialize + CanonicalDeserialize,")]
-pub struct MerkleProof<E, I, T, const ARITY: usize>
-where
-    E: Element,
-    I: Index,
-    T: NodeValue,
-{
-    /// Proof of inclusion for element at index `pos`
-    #[serde(with = "canonical")]
-    pub pos: I,
-    /// Nodes of proof path, from root to leaf
-    pub proof: MerklePath<E, I, T>,
+/// A (non)membership Merkle proof consists of all values of siblings of a
+/// Merkle path.
+#[derive(
+    Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, CanonicalSerialize, CanonicalDeserialize,
+)]
+#[tagged("MERKLE_PROOF")]
+pub struct MerkleTreeProof<T: NodeValue>(pub Vec<Vec<T>>);
+
+impl<T: NodeValue> super::MerkleProof<T> for MerkleTreeProof<T> {
+    /// Expected height of the Merkle tree.
+    fn height(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Return all values of siblings of this Merkle path
+    fn path_values(&self) -> &[Vec<T>] {
+        &self.0
+    }
 }
 
-impl<E, I, T, const ARITY: usize> MerkleProof<E, I, T, ARITY>
+/// Verify a merkle proof
+/// * `commitment` - a merkle tree commitment
+/// * `pos` - zero-based index of the leaf in the tree
+/// * `element` - the leaf value, None if verifying a non-membership proof
+/// * `proof` - a membership proof for `element` at given `pos`
+/// * `returns` - Ok(true) if the proof is accepted, Ok(false) if not. Err() if
+///   the proof is not well structured, E.g. not for this merkle tree.
+pub(crate) fn verify_merkle_proof<E, H, I, const ARITY: usize, T>(
+    commitment: &MerkleTreeCommitment<T>,
+    pos: &I,
+    element: Option<&E>,
+    proof: &[Vec<T>],
+) -> Result<VerificationResult, MerkleTreeError>
 where
     E: Element,
-    I: Index,
+    I: Index + ToTraversalPath<ARITY>,
     T: NodeValue,
+    H: DigestAlgorithm<E, I, T>,
 {
-    /// Return the height of this proof.
-    pub fn tree_height(&self) -> usize {
-        self.proof.len()
-    }
-
-    /// Form a `MerkleProof` from a given index and Merkle path.
-    pub fn new(pos: I, proof: MerklePath<E, I, T>) -> Self {
-        MerkleProof { pos, proof }
-    }
-
-    /// Return the index of this `MerkleProof`.
-    pub fn index(&self) -> &I {
-        &self.pos
-    }
-
-    /// Return the element associated with this `MerkleProof`. None if it's a
-    /// non-membership proof.
-    pub fn elem(&self) -> Option<&E> {
-        match self.proof.first() {
-            Some(MerkleNode::Leaf { elem, .. }) => Some(elem),
-            _ => None,
+    if proof.len() == commitment.height() {
```

Comment:
I mean the height check may not be necessary now.

Reply:
My understanding of your comments is: now that the caller supplies the element data, there is no need to check tree height because such a check is necessary only if we assume the caller can find hash collisions. This makes sense to me. We should delete the check.

In fact, I see no reason why `commitment` needs to be a `MerkleTreeCommitment`. Really all we need is the hash digest, so we should take only that as an arg instead of `MerkleTreeCommitment`.

Reply:
done in [e2e5c33](https://github.com/EspressoSystems/jellyfish/pull/692/commits/e2e5c3377fe094cf2dc84dd004ea1b53a6a9e7f4)

Reply:
wow so much deleted code! 🎉 🥳